tests/test_ssds.py

```diff
@@ -50,23 +52,32 @@ def tearDownClass(cls):
         ssds.blobstore.AWS_MIN_CHUNK_SIZE = cls._old_aws_min_chunk_size
 
     def test_upload(self):
+        threads = 2
         submission_name = "this_is_a_test_submission"
         tests = [
-            ("aws sync", S3_SSDS, f"{uuid4()}", None),
-            ("aws async", S3_SSDS, f"{uuid4()}", 4),
-            ("gcp sync", GS_SSDS, f"{uuid4()}", None),
-            ("gcp async", GS_SSDS, f"{uuid4()}", 4),
+            ("local to aws", LocalBlobStore, self.testdir, "", S3_SSDS, f"{uuid4()}"),
+            ("local to gcp", LocalBlobStore, self.testdir, "", GS_SSDS, f"{uuid4()}"),
+            ("aws to aws", S3BlobStore, "org-hpp-ssds-upload-test", f"{self.multifile_pfx}", S3_SSDS, f"{uuid4()}"),
+            ("aws to gcp", S3BlobStore, "org-hpp-ssds-upload-test", f"{self.multifile_pfx}", GS_SSDS, f"{uuid4()}"),
+            ("gcp to aws", GSBlobStore, "org-hpp-ssds-upload-test", f"{self.multifile_pfx}", S3_SSDS, f"{uuid4()}"),
+            ("gcp to gcp", GSBlobStore, "org-hpp-ssds-upload-test", f"{self.multifile_pfx}", GS_SSDS, f"{uuid4()}"),
         ]
-        for test_name, ds, submission_id, threads in tests:
-            with self.subTest(test_name):
+        for test_name, src_blobstore, src_bucket, src_pfx, dst_ds, submission_id in tests:
+            src_url = f"{src_blobstore.schema}{src_bucket}/{src_pfx}"
+            with self.subTest(test_name, threads=threads):
                 start_time = time.time()
-                for ssds_key in ds.upload(self.testdir, submission_id, submission_name, threads):
+                for ssds_key in dst_ds.upload(src_url, submission_id, submission_name, threads):
                     pass
-                print(f"{test_name} upload duration:", time.time() - start_time)
-                for blob in LocalBlobStore(self.testdir).list():
-                    expected_ssds_key = ds._compose_ssds_key(submission_id, submission_name, blob.key)
-                    dst_key = f"{ds.prefix}/{expected_ssds_key}"
-                    self.assertEqual(blob.size(), ds.blobstore.blob(dst_key).size())
+                print(test_name, "took", time.time() - start_time, "seconds")
+                for blob in src_blobstore(src_bucket).list(src_pfx):
+                    subdir = blob.key.replace(src_pfx, "")
```

Comment:
Will this be a problem with something like `gs://bucket/cram_inputs/cram_inputs/cram_inputs/cram.txt`?

Maybe only replace the first?  For example:

```
>>> text = 'some kind of something'.replace("some", "", 1)
>>> print(text)
 kind of something
```

Reply:
will fix